src/go/wsl-helper/pkg/dockerproxy/util/reverse_proxy.go

```diff
@@ -0,0 +1,277 @@
+	// Create a new context with cancellation to ensure we can stop the flush
+	// The context will be canceled when the request is done or if needed earlier
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Establish a connection to the backend using a custom Dial method
+	backendConn, err := proxy.Dial("", "")
+	if err != nil {
+		proxy.logf("failed to connect to the backend: %v", err.Error())
+		http.Error(w, err.Error(), http.StatusBadGateway)
+		return
+	}
+	defer backendConn.Close()
+
+	// Create a new HTTP request with the same headers
+	url := targetProtocol + hostHeaderValue + r.RequestURI
+	newReq, err := http.NewRequestWithContext(ctx, r.Method, url, r.Body)
+	if err != nil {
+		proxy.logf("failed to create a request for the backend: %v", err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	newReq.Header = r.Header
+
+	// Director function
+	// Allows complete customization of the outgoing request
+	if proxy.Director != nil {
+		proxy.Director(newReq)
+	}
+	// Prevent automatic connection closure
+	newReq.Close = false
+
+	// Forward the modified request to the backend
+	if err = newReq.Write(backendConn); err != nil {
+		proxy.logf("failed to forward the request to the backend: %v", err.Error())
+		http.Error(w, err.Error(), http.StatusBadGateway)
+		return
+	}
+
+	// Read the response from the backend
+	backendResponse, err := http.ReadResponse(bufio.NewReader(backendConn), newReq)
+	if err != nil {
+		proxy.logf("failed to read the response from the backend: %v", err.Error())
+		http.Error(w, err.Error(), http.StatusBadGateway)
+		return
+	}
+	defer backendResponse.Body.Close()
+
+	// ModifyResponse function
+	// Allows post-processing of the backend response
+	if proxy.ModifyResponse != nil {
+		err := proxy.ModifyResponse(backendResponse)
+		if err != nil {
+			proxy.logf("failed to modify the response from the backend: %v", err.Error())
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	// Propagate backend response headers to the client
+	for key, values := range backendResponse.Header {
+		for _, value := range values {
+			w.Header().Add(key, value)
+		}
+	}
+
+	// Write the response status code and headers and flush it immediately
+	w.WriteHeader(backendResponse.StatusCode)
+	flusher, ok := w.(http.Flusher)
+	if !ok {
+		panic("expected http.ResponseWriter to be an http.Flusher")
+	}
+	flusher.Flush()
+
+	// Check if the response has a status code of 101 (Switching Protocols)
+	if backendResponse.StatusCode == http.StatusSwitchingProtocols {
+		proxy.handleUpgradedConnection(w, backendConn)
+		return
+	}
+
+	// Start periodic flushing in a background goroutine
+	// Supports long-running, streaming responses
+	go periodicHttpFlush(w, ctx)
+
+	// Stream the response body back to the client
+	_, err = io.Copy(w, backendResponse.Body)
+	if err != nil {
+		proxy.logf("failed to stream the response body to the client: %v", err.Error())
+		return
```

Comment:
I'm surprised Linter didn't get mad over this return 😄  because your function would have returned regardless. Also, this is totally optional but you could simplify this to something like: 
```
	// Stream the response body back to the client
	if _, err = io.Copy(w, backendResponse.Body); err != nil {
		proxy.logf("failed to stream the response body to the client: %v", err.Error())
	}
```